File: computing/interview/design-patterns/structural/facade/java/FacadeDemo.java
package interview.designpatterns.structural.facade;

final class Amplifier {
    void on() {
        System.out.println("Amplifier on");
    }

    void setVolume(int level) {
        System.out.printf("Amplifier volume %d%n", level);
    }
}

final class StreamingPlayer {
    void on() {
        System.out.println("Player on");
    }

    void playMovie(String title) {
        System.out.printf("Now playing: %s%n", title);
    }

    void stop() {
        System.out.println("Stopping playback");
    }
}

final class Projector {
    void on() {
        System.out.println("Projector on");
    }

    void wideScreenMode() {
        System.out.println("Setting widescreen mode");
    }
}

final class HomeTheaterFacade {
    private final Amplifier amp;
    private final StreamingPlayer player;
    private final Projector projector;

    HomeTheaterFacade(Amplifier amp, StreamingPlayer player, Projector projector) {
        this.amp = amp;
        this.player = player;
        this.projector = projector;
    }

    void watchMovie(String title) {
        System.out.println("Get ready to watch a movie...");
        amp.on();
        amp.setVolume(7);
        projector.on();
        projector.wideScreenMode();
        player.on();
        player.playMovie(title);
    }

    void endMovie() {
        System.out.println("Shutting movie theater down...");
        player.stop();
    }
}

public final class FacadeDemo {
    private FacadeDemo() {}

    public static void main(String[] args) {
        HomeTheaterFacade theater = new HomeTheaterFacade(new Amplifier(), new StreamingPlayer(), new Projector());
        theater.watchMovie("Inception");
        theater.endMovie();
    }
}

File: computing/interview/design-patterns/structural/facade/csharp/FacadeDemo.cs
using System;

namespace Interview.DesignPatterns.Structural.Facade
{
    public sealed class Amplifier
    {
        public void On() => Console.WriteLine("Amplifier on");
        public void SetVolume(int level) => Console.WriteLine($"Amplifier volume {level}");
    }

    public sealed class StreamingPlayer
    {
        public void On() => Console.WriteLine("Player on");
        public void PlayMovie(string title) => Console.WriteLine($"Now playing: {title}");
        public void Stop() => Console.WriteLine("Stopping playback");
    }

    public sealed class Projector
    {
        public void On() => Console.WriteLine("Projector on");
        public void WideScreenMode() => Console.WriteLine("Setting widescreen mode");
    }

    public sealed class HomeTheaterFacade
    {
        private readonly Amplifier _amp;
        private readonly StreamingPlayer _player;
        private readonly Projector _projector;

        public HomeTheaterFacade(Amplifier amp, StreamingPlayer player, Projector projector)
        {
            _amp = amp;
            _player = player;
            _projector = projector;
        }

        public void WatchMovie(string title)
        {
            Console.WriteLine("Get ready to watch a movie...");
            _amp.On();
            _amp.SetVolume(7);
            _projector.On();
            _projector.WideScreenMode();
            _player.On();
            _player.PlayMovie(title);
        }

        public void EndMovie()
        {
            Console.WriteLine("Shutting movie theater down...");
            _player.Stop();
        }
    }

    public static class Demo
    {
        public static void Main()
        {
            var theater = new HomeTheaterFacade(new Amplifier(), new StreamingPlayer(), new Projector());
            theater.WatchMovie("Inception");
            theater.EndMovie();
        }
    }
}
